gateway/src/apicast/policy/3scale_batcher/3scale_batcher.lua

```diff
@@ -0,0 +1,160 @@
+local default_batch_reports_seconds = 10
+
+local _M = policy.new('Caching policy')
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new(config)
+
+  local auths_ttl = config.auths_ttl or default_auths_ttl
+  self.auths_cache = AuthsCache.new(ngx.shared.cached_auths, auths_ttl)
+
+  self.reports_batcher = ReportsBatcher.new(
+    ngx.shared.batched_reports, 'batched_reports_locks')
+
+  self.batch_reports_seconds = config.batch_reports_seconds or
+                               default_batch_reports_seconds
+
+  self.report_timer_on = false
+
+  local semaphore_report_timer, err = semaphore.new(1)
+  if not semaphore_report_timer then
+    ngx.log(ngx.ERR, "Create semaphore failed: ", err)
+  end
+  self.semaphore_report_timer = semaphore_report_timer
+
+  return self
+end
+
+-- Converts a usage to the format expected by the 3scale backend client.
+local function format_usage(usage)
+  local res = {}
+
+  local usage_metrics = usage.metrics
+  local usage_deltas = usage.deltas
+
+  for _, metric in ipairs(usage_metrics) do
+    local delta = usage_deltas[metric]
+    res['usage[' .. metric .. ']'] = delta
+  end
+
+  return res
+end
+
+local function set_flags_to_avoid_auths_in_apicast(context)
+  context.skip_apicast_access = true
+  context.skip_apicast_post_action = true
+end
+
+local function report(_, service_id, backend, reports_batcher)
+  local reports = reports_batcher:get_all(service_id)
+  reporter.report(reports, service_id, backend, reports_batcher)
+end
+
+-- This starts a timer on each worker.
+-- Starting a timer on each worker means that there will be more calls to
+-- 3scale backend, and the config param 'batch_report_seconds' becomes
+-- more confusing because the reporting frequency will be affected by the
+-- number of APIcast workers.
+-- If we started a timer just on one of the workers, it could die, and then,
+-- there would not be any reporting.
+local function ensure_report_timer_on(self, service_id, backend)
+  local check_timer = self.semaphore_report_timer:wait(0)
+
+  if check_timer then
+    if not self.report_timer_on then
+      ngx.timer.every(self.batch_reports_seconds, report,
+        service_id, backend, self.reports_batcher)
+
+      self.report_timer_on = true
+    end
+
+    self.semaphore_report_timer:post()
+  end
+end
+
+local function rejection_reason_from_headers(response_headers)
+  return response_headers and response_headers['3scale-rejection-reason']
+end
+
+local function error(service, rejection_reason)
+  if rejection_reason == 'limits_exceeded' then
+    return errors.limits_exceeded(service)
+  else
+    return errors.authorization_failed(service)
+  end
+end
+
+function _M:access(context)
+  local backend = backend_client:new(context.service, http_ng_resty)
+  local usage = context.usage
+  local service = context.service
+  local service_id = service.id
+  local credentials = context.credentials
+
+  ensure_report_timer_on(self, service_id, backend)
```

Comment:
We definitely have to expose some way to init/destroy policies so they can do stuff like this. Global policies can do this in init_worker and we need some mechanics for local policies too. Including destructors because they can get GC'd.